### DISCUSSION TITLE
when using a page decorator and functional pages, the save detection doesn't work

### DISCUSSION THREAD

**Discussion by vickorian** [zauberzeug/nicegui#3071] [Q&A] [answered]
### Question

When just using standard ui.run() and not separating the code with pages via @ui.page decorator, when i make a save to the code, it automatically updates the page allowing for quick iteration. But when the decorator is in place and the code is running under a functional, this behavior doesn't occur and requires one to close the running page, and re-run the code, dramatically increasing dev time. 

Is this the intended behavior or is there a way to allow saving to auto-recreate the page?

Thanks.

**Comment by rodja** [comment]
No, that should not happen. And it does work for me (and many others). Please share a minimal reproduction of the code which produces this behaviour.

**Reply by vickorian** [reply]
@rodja 

So something just this simple will throw it off. When running VS code, making changes and saving will not push the changes, but removing the ui.page('/') and the async def main(): will lead to the correct behavior. When saving, you'll get the 
"WatchFiles detected changes in 'Popout_test.py'. Reloading..." but the uvicorn server will never reload.

Do you need to put `if __name__ == "__main__":` and the `ui.run()` after that?

```py
@ui.page('/')
async def main():
    async def replace():
        print('hello')

    async def print_date_val():
        print("Today")


    ui.button('dog', on_click=print_date_val)
    ui.button('Replace', on_click=replace)

ui.run(host="127.0.0.1")
```

**Reply by rodja** [reply]
Sorry, your example reloads just fine when I edit/save. Could you try a docker container, new pyenv or similar?

**Reply by vickorian** [reply, marked as answer]
@rodja 

Not sure what was causing the issue, but i'm pretty sure it was an outdated version of click. I'd been having this problem for a while, but when i updated to the most recent version of nicegui i noticed that i was getting a warning on PIP that click was outdated and as a result PIP was having trouble resolving the environment of certain packages like sphinx. The new version of Nicegui installed, but the problem persisted. Updated click, re-installed the most recent nicegui and got no warnings regarding unresolved environment. Problem resolved itself after updating click.

Maybe make a note of it in requirements.txt?

**Reply by falkoschindler** [reply]
Great! 🙂
The click package seems to be a uvicorn dependency. So we guess it's a uvicorn issue, which might already be tracked in their repo.

**Reply by vickorian** [reply]
That would make sense. The dependency tree must be relatively deep for all subpackages so things like this are bound to happen.
